src/bin/server/transcoder/actor.rs

```diff
@@ -0,0 +1,215 @@
+        Transcoder {
+            disk_cache: Rc::new(RefCell::new(cache)),
+            future_cache: Rc::new(RefCell::new(HashMap::new())),
+            temp,
+        }
+    }
+
+    /// Transcodes the file requested by the message into a temporary path.
+    fn transcode<P: AsRef<Path>>(
+        &self,
+        msg: &TranscodeMessage<P>,
+    ) -> Box<Future<Item = (Output, PathBuf), Error = io::Error>> {
+        let temporary_file_path = self.temp.get_file();
+
+        // TODO: Configurable FFMpeg Instance
+        // TODO: Handle Non Zero Exit Code
+        Box::new(
+            Command::new("ffmpeg")
+                .args(msg.get_ffmpeg_args(&temporary_file_path))
+                .output_async()
+                .map(move |output| (output, temporary_file_path.to_path_buf())),
+        )
+    }
+
+    /// Transcodes the file requested by the message, updating the relevant caches.
+    fn transcode_and_cache<P: AsRef<Path>>(
+        &self,
+        msg: &TranscodeMessage<P>,
+    ) -> ResponseFuture<(), io::Error> {
+        let key = msg.to_key();
+        let future_map_key = key.clone();
+
+        let cache = self.disk_cache.clone();
+        let future_cache = self.future_cache.clone();
+
+        let transcode_future = self
+            .transcode(&msg)
+            .map(move |(_output, temporary_file_path)| {
+                // TODO: Try Using Failure Errors
+                // Result ignored here because if inserting a file fails later,
+                // getting it out of the cache will fail also. It's better to
+                // start the error here, but forwarding errors from here is
+                // difficult because Shared<Future<...>> turns the error into
+                // an error reference. error-chain errors can't be cloned.
+                let _ = cache
+                    .borrow_mut()
+                    .insert_file(&future_map_key, temporary_file_path);
+
+                future_cache.borrow_mut().remove(&future_map_key);
+
+                ()
+            });
+
+        let boxed_transcode_future: Box<Future<Item = (), Error = io::Error>> =
+            Box::new(transcode_future);
+
+        let shared_future = boxed_transcode_future.shared();
+
+        self.future_cache
+            .borrow_mut()
+            .insert(key.clone(), shared_future.clone());
+
+        Box::new(shared_future.map(|_| ()).map_err(|err| err.kind().into()))
+    }
+
+    fn disk_cache_has(&self, key: &OsStr) -> bool {
+        self.disk_cache.borrow().contains_key(key)
+    }
+
+    fn future_cache_has(&self, key: &OsStr) -> bool {
+        self.future_cache.borrow().contains_key(key)
+    }
+
+    /// Returns a future which resolves when the transcode job associated with the message is
+    /// complete and the transcoded file is in the disk_cache.
+    fn get_cache_populated_future<P: AsRef<Path>>(
+        &self,
+        msg: &TranscodeMessage<P>,
+    ) -> ResponseFuture<(), io::Error> {
```

Comment:
This method does not "get" anything, so the name should be changed to reflect what it actually does (beyond "it gets a future").

Reply:
Please suggest a better name. I'm not sure how to describe the idea that this function checks two caches returning from them or doing a computation, populating the cache and returning the result. Maybe `memoized_cache_populated_future` or something?